Enable @FormParam for EntityParts

jim-krueger opened the pull request:
Fixes #25017

jim-krueger committed:
Enable @FormParam for EntityParts

jim-krueger commented:
#libby #build

jim-krueger commented:
#libby #build

jim-krueger commented:
#build

jim-krueger commented:
#build #spawn.fullfat.buckets=com.ibm.ws.jaxrs.2.0_fat,com.ibm.ws.jaxrs.2.0.cdi.1.2_fat,com.ibm.ws.jaxrs.2.0.client_fat,com.ibm.ws.jaxrs.2.1_fat,com.ibm.ws.jaxrs.2.1_fat_executor,com.ibm.ws.jaxrs.2.1_fat_extended,com.ibm.ws.jaxrs.2.1.cdi.2.0_fat,com.ibm.ws.jaxrs.2.1.client_fat,com.ibm.ws.jaxrs.2.1.sse_fat,com.ibm.ws.jaxrs.2.x_fat_clientProps

jim-krueger committed:
Replace multipart test

jim-krueger commented:
#build #spawn.fullfat.buckets=com.ibm.ws.jaxrs.2.0_fat,com.ibm.ws.jaxrs.2.0.cdi.1.2_fat,com.ibm.ws.jaxrs.2.0.client_fat,com.ibm.ws.jaxrs.2.1_fat,com.ibm.ws.jaxrs.2.1_fat_executor,com.ibm.ws.jaxrs.2.1_fat_extended,com.ibm.ws.jaxrs.2.1.cdi.2.0_fat,com.ibm.ws.jaxrs.2.1.client_fat,com.ibm.ws.jaxrs.2.1.sse_fat,com.ibm.ws.jaxrs.2.x_fat_clientProps

jim-krueger committed:
Cleanup

jim-krueger commented:
#libby

jim-krueger commented:
#libby

jim-krueger commented:
#libby #build #spawn.fullfat.buckets=com.ibm.ws.jaxrs.2.0_fat,com.ibm.ws.jaxrs.2.0.cdi.1.2_fat,com.ibm.ws.jaxrs.2.0.client_fat,com.ibm.ws.jaxrs.2.1_fat,com.ibm.ws.jaxrs.2.1_fat_executor,com.ibm.ws.jaxrs.2.1_fat_extended,com.ibm.ws.jaxrs.2.1.cdi.2.0_fat,com.ibm.ws.jaxrs.2.1.client_fat,com.ibm.ws.jaxrs.2.1.sse_fat,com.ibm.ws.jaxrs.2.x_fat_clientProps,io.openliberty.restfulWS.3.0_fat,io.openliberty.restfulWS.3.0.cdi.3.0_fat,io.openliberty.restfulWS.3.0.client_fat,io.openliberty.restfulWS.3.1_fat

jhanders34 commented on dev/io.openliberty.org.jboss.resteasy.common.ee10/src/org/jboss/resteasy/core/FormParamInjector.java:
```diff
@@ -68,6 +73,8 @@ public FormParamInjector(final Class type, final Type genericType, final Accessi
    @Override
    public Object inject(HttpRequest request, HttpResponse response, boolean unwrapAsync)
    {
+       boolean debug = tc.isDebugEnabled() && TraceComponent.isAnyTracingEnabled();  //Liberty Change
```
```suggestion
       boolean debug = TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled();  //Liberty Change
```

jim-krueger replied:
Done

jhanders34 commented on dev/io.openliberty.org.jboss.resteasy.common.ee10/src/org/jboss/resteasy/core/FormParamInjector.java:
```diff
@@ -68,6 +73,8 @@ public FormParamInjector(final Class type, final Type genericType, final Accessi
    public Object inject(HttpRequest request, HttpResponse response, boolean unwrapAsync)
    {
+       boolean debug = tc.isDebugEnabled() && TraceComponent.isAnyTracingEnabled();  //Liberty Change
+       List<String> list = null;
```
This should be lower after the EntityPart, List<EntityPart> and InputStream checks.

jim-krueger replied:
Done

jhanders34 commented on dev/io.openliberty.org.jboss.resteasy.server.ee10/src/io/openliberty/org/jboss/resteasy/server/component/RESTfulServletContainerInitializer.java:
```diff
@@ -17,6 +17,16 @@
 import java.util.Map;
 import java.util.Set;
 
+//import org.jboss.resteasy.plugins.server.servlet.HttpServlet30Dispatcher;
```
```suggestion
```

jim-krueger replied:
Done.

jhanders34 commented on dev/io.openliberty.restfulWS.3.1_fat/fat/src/io/openliberty/restfulWS31/fat/multipart/MultipartResource.java:
```diff
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (c) 2022 IBM Corporation and others.
+ * Copyright (c) 2023 IBM Corporation and others.
```
```suggestion
 * Copyright (c) 2022, 2023 IBM Corporation and others.
```

jim-krueger replied:
Done.

jim-krueger commented:
#libby #build #spawn.fullfat.buckets=com.ibm.ws.jaxrs.2.0_fat,com.ibm.ws.jaxrs.2.0.cdi.1.2_fat,com.ibm.ws.jaxrs.2.0.client_fat,com.ibm.ws.jaxrs.2.1_fat,com.ibm.ws.jaxrs.2.1_fat_executor,com.ibm.ws.jaxrs.2.1_fat_extended,com.ibm.ws.jaxrs.2.1.cdi.2.0_fat,com.ibm.ws.jaxrs.2.1.client_fat,com.ibm.ws.jaxrs.2.1.sse_fat,com.ibm.ws.jaxrs.2.x_fat_clientProps,io.openliberty.restfulWS.3.0_fat,io.openliberty.restfulWS.3.0.cdi.3.0_fat,io.openliberty.restfulWS.3.0.client_fat,io.openliberty.restfulWS.3.1_fat

jhanders34 commented:
You also should add a FormParam test with a mix of EntityPart and Strings.  Do you need more `@PUT` and `@GET` tests as well?

jim-krueger committed:
Code review comments

jim-krueger committed:
Added tests and code review comments.

jim-krueger commented:
#libby #build #spawn.fullfat.buckets=com.ibm.ws.jaxrs.2.0_fat,com.ibm.ws.jaxrs.2.0.cdi.1.2_fat,com.ibm.ws.jaxrs.2.0.client_fat,com.ibm.ws.jaxrs.2.1_fat,com.ibm.ws.jaxrs.2.1_fat_executor,com.ibm.ws.jaxrs.2.1_fat_extended,com.ibm.ws.jaxrs.2.1.cdi.2.0_fat,com.ibm.ws.jaxrs.2.1.client_fat,com.ibm.ws.jaxrs.2.1.sse_fat,com.ibm.ws.jaxrs.2.x_fat_clientProps,io.openliberty.restfulWS.3.0_fat,io.openliberty.restfulWS.3.0.cdi.3.0_fat,io.openliberty.restfulWS.3.0.client_fat,io.openliberty.restfulWS.3.1_fat

WhiteCat22 reviewed (dismissed):
I didn't walk through all the tests, but the product changes make sense to me

jim-krueger committed:
Assign default config values

jim-krueger commented:
#libby #build #spawn.fullfat.buckets=com.ibm.ws.jaxrs.2.0_fat,com.ibm.ws.jaxrs.2.0.cdi.1.2_fat,com.ibm.ws.jaxrs.2.0.client_fat,com.ibm.ws.jaxrs.2.1_fat,com.ibm.ws.jaxrs.2.1_fat_executor,com.ibm.ws.jaxrs.2.1_fat_extended,com.ibm.ws.jaxrs.2.1.cdi.2.0_fat,com.ibm.ws.jaxrs.2.1.client_fat,com.ibm.ws.jaxrs.2.1.sse_fat,com.ibm.ws.jaxrs.2.x_fat_clientProps,io.openliberty.restfulWS.3.0_fat,io.openliberty.restfulWS.3.0.cdi.3.0_fat,io.openliberty.restfulWS.3.0.client_fat,io.openliberty.restfulWS.3.1_fat

jim-krueger commented:
Added logic to FormParamInjector to account for @FormParam multipart parameters on resource methods (specifically EntityPart, String, and InputStream per the specification).

In addition:  For some aspects of multipart processing, RESTEasy appears to require the @MultipartConfig annotation on a servlet.   For Liberty that translates to adding that annotation to our IBMRestServlet.   This ended up causing some NPEs in beanvalidation FAT until I added default values (replacing null) to config in WebAppConfiguratorHelper.   The addition of the annotation may be something we will want to remove at some point in the future.

jhanders34 commented on dev/io.openliberty.restfulWS.3.1_fat/fat/src/io/openliberty/restfulWS31/fat/multipart/MultipartTestServlet.java:
```diff
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (c) 2022 IBM Corporation and others.
+ * Copyright (c) 2023 IBM Corporation and others.
```
```suggestion
 * Copyright (c) 2022, 2023 IBM Corporation and others.
```

jim-krueger committed:
Update dev/io.openliberty.restfulWS.3.1_fat/fat/src/io/openliberty/restfulWS31/fat/multipart/MultipartTestServlet.java

Code review comments

Co-authored-by: Jared Anderson <jhanders@us.ibm.com>

jim-krueger commented:
#libby

jim-krueger commented:
#run-libby-bot
